File: csharp/test/Plugins.cs
// RUN: make test-plugins
// RUN-SOME: ./test/bin/Debug/net8.0/sekretotest plugins
//
// THE PLUGIN SEAM, from both sides.
//
// Moving the provider kinds that open sockets and spawn processes out of
// the core made a consumer's PLUGIN LIST load-bearing: a kind nobody
// passed in is not in the catalog, and a chain naming it is refused. That
// is the intended behaviour, and it means a consumer can be broken
// without a single conformance check noticing - the conformance suite
// passes every plugin to every chain it builds, so it can never see a
// missing one. So the full set is pinned here: it holds every kind, every
// kind builds, and the CLI passes it.
//
// And from the other side: the core assembly must not be able to REACH a
// plugin. In csharp that is not a convention, it is the assembly
// reference graph, and the last three cases read it out of the compiled
// artifact.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using Voxgig.Plugin;
using Voxgig.Sekreto;
using Voxgig.Sekreto.Plugins;

internal static class Seam
{
    private static readonly string[] PLUGINS =
    {
        "awsparams", "awssecrets", "azuresecrets", "boru", "doppler", "gcpsecrets",
        "hashicorp", "infisical", "onepassword", "secretspec",
    };

    // Every kind this library ships, built in or plugged in, in one sorted
    // list - which is also the order a chain of all of them reports.
    private static readonly string[] EVERY =
        new[] { "dotenv", "env", "file", "memory" }.Concat(PLUGINS).OrderBy(n => n, StringComparer.Ordinal).ToArray();

    // --- assertions -------------------------------------------------------

    private static void Eq(object got, object want, string what)
    {
        string gottext = Show(got);
        string wanttext = Show(want);

        if (gottext != wanttext)
        {
            throw new Exception(what + ":\n  got  " + gottext + "\n  want " + wanttext);
        }
    }

    private static string Show(object value)
    {
        if (value is IEnumerable<string> texts)
        {
            return "[" + string.Join(", ", texts) + "]";
        }

        if (value is IEnumerable<object> items)
        {
            return "[" + string.Join(", ", items.Select(Show)) + "]";
        }

        return null == value ? "<null>" : Convert.ToString(value);
    }

    private static void True(bool got, string what)
    {
        if (!got)
        {
            throw new Exception(what);
        }
    }

    private static Exception Caught(Action body, string what)
    {
        try
        {
            body();
        }
        catch (Exception err)
        {
            return err;
        }

        throw new Exception(what + ": nothing was thrown");
    }

    private static string Refused(Action body, string what)
    {
        Exception err = Caught(body, what);

        if (!(err is SekretoError))
        {
            throw new Exception(what + ": " + err.GetType().Name + " - " + err.Message);
        }

        return err.Message;
    }

    // --- helpers ----------------------------------------------------------

    private static Dictionary<string, object> Spec(params object[] pairs)
    {
        var out_ = new Dictionary<string, object>();

        for (int index = 0; index + 1 < pairs.Length; index += 2)
        {
            out_[Convert.ToString(pairs[index])] = pairs[index + 1];
        }

        return out_;
    }

    private static List<string> Names(IEnumerable<Definition> definitions)
    {
        return definitions.Select(d => d.Name).ToList();
    }

    private static string Here(params string[] parts)
    {
        string dir = Path.GetDirectoryName(typeof(Seam).Assembly.Location);

        for (int step = 0; step < 8 && null != dir; step++)
        {
            string cand = Path.Combine(new[] { dir }.Concat(parts).ToArray());

            if (File.Exists(cand))
            {
                return cand;
            }

            dir = Path.GetDirectoryName(dir);
        }

        throw new Exception("seam: file not found: " + string.Join("/", parts));
    }

    /// <summary>A provider kind a consumer wrote, not this library.</summary>
    private sealed class Shouty : IProvider
    {
        private readonly Dictionary<string, object> values;

        internal Shouty(Dictionary<string, object> values)
        {
            this.values = values ?? new Dictionary<string, object>();
        }

        public string Lookup(string name)
        {
            return values.TryGetValue(name.ToUpperInvariant(), out object found)
                ? Convert.ToString(found)
                : null;
        }

        public string Describe()
        {
            return "shouty";
        }
    }

    private sealed class Replaced : IProvider
    {
        public string Lookup(string name)
        {
            return "replaced";
        }

        public string Describe()
        {
            return "memory";
        }
    }

    // --- the cases --------------------------------------------------------

    internal static List<KeyValuePair<string, Action>> Cases()
    {
        var cases = new List<KeyValuePair<string, Action>>();

        void Case(string name, Action body)
        {
            cases.Add(new KeyValuePair<string, Action>(name, body));
        }

        Case("the full set holds every kind", () =>
        {
            Eq(Names(SekretoPlugins.All()).OrderBy(n => n, StringComparer.Ordinal), PLUGINS,
                "SekretoPlugins.All()");
            Eq(Names(Providers.Builtins()), Providers.BuiltinKinds, "Providers.Builtins()");
            Eq(Providers.PluginKinds.OrderBy(n => n, StringComparer.Ordinal), PLUGINS,
                "Providers.PluginKinds");
        });

        // Naming a kind is not enough: a kind can be in the catalog and
        // still fail to build. Construction is what the CLI does before any
        // network.
        Case("every kind builds from a spec", () =>
        {
            var chain = new List<object>();

            foreach (string kind in EVERY)
            {
                chain.Add(Spec(
                    "kind", kind, "addr", "http://127.0.0.1:8200", "token", "t",
                    "dir", "/tmp", "file", "/tmp/.env", "values", new Dictionary<string, object>()));
            }

            var secrets = new Sekreto(new SekretoOptions
            {
                Providers = chain, Plugins = SekretoPlugins.All(), Cache = false,
            });

            Eq(secrets.Stores(), EVERY, "stores");
            Eq(secrets.Host.List().Keys, EVERY, "host refs");
            Eq(secrets.Host.List().Values.Distinct(), new[] { "live" }, "host statuses");
        });

        Case("the cli passes the full set", () =>
        {
            string source = File.ReadAllText(Here("cli", "Program.cs"));
            True(source.Contains("using Voxgig.Sekreto.Plugins;"), "the CLI names the plugin assembly");
            True(source.Contains("Plugins = SekretoPlugins.All(),"), "the CLI passes the full set");

            string project = File.ReadAllText(Here("cli", "SekretoCli.csproj"));
            True(project.Contains("../plugins/SekretoPlugins.csproj"), "the CLI references the plugin assembly");
        });

        // --- what a consumer sees -----------------------------------------

        Case("one plugin is enough for a chain that names only it", () =>
        {
            var secrets = new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { Hashicorp.Plugin },
                Providers = new List<object>
                {
                    Spec("kind", "memory", "values", Spec("API_TOKEN", "tok01")),
                    Spec("kind", "hashicorp", "name", "prod",
                         "addr", "https://vault.example.com", "token", "t"),
                },
            });

            Eq(secrets.Stores(), new[] { "memory", "prod" }, "stores");
            Eq(secrets.Sources(),
                new[] { "memory", "hashicorp:https://vault.example.com/secret" }, "sources");
            Eq(secrets.Get("api.token"), "tok01", "get");

            // The plugin host is what the chain is made of, and it reads
            // like the chain: the kind, or kind$store for a named store.
            Eq(secrets.Host.List().Keys, new[] { "hashicorp$prod", "memory" }, "host refs");
            Eq(secrets.Catalog.Names(),
                new[] { "dotenv", "env", "file", "hashicorp", "memory" }, "catalog");
        });

        Case("a kind that was not passed in is refused, naming the fix", () =>
        {
            Eq(Refused(() => new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { Hashicorp.Plugin },
                Providers = new List<object> { Spec("kind", "doppler", "token", "t") },
            }), "an unloaded plugin kind"),
                "sekreto: unknown provider kind: doppler"
                + " (available: dotenv, env, file, hashicorp, memory)"
                + " - doppler is a sekreto plugin, not built in: pass it in the plugins option",
                "the message names the fix");

            // A kind nobody ships is a typo, and gets no such hint.
            Eq(Refused(() => new Sekreto(new SekretoOptions
            {
                Providers = new List<object> { Spec("kind", "vualt") },
            }), "a kind nobody ships"),
                "sekreto: unknown provider kind: vualt (available: dotenv, env, file, memory)",
                "a typo gets no hint");
        });

        // Two providers MAY share a store name - a directed read walks both,
        // and the spec pins it - but an instance ref may not, so the second
        // gets a numbered tag from the host and keeps its store name.
        Case("a repeated store name keeps the store and numbers the instance", () =>
        {
            var secrets = new Sekreto(new SekretoOptions
            {
                Providers = new List<object>
                {
                    Spec("kind", "memory", "values", Spec()),
                    Spec("kind", "memory", "values", Spec("API_TOKEN", "second")),
                    Spec("kind", "memory", "name", "pair", "values", Spec()),
                    Spec("kind", "memory", "name", "pair", "values", Spec("API_TOKEN", "pair2")),
                },
            });

            Eq(secrets.Stores(), new[] { "memory", "pair" }, "stores");
            Eq(secrets.Host.List().Keys,
                new[] { "memory", "memory$1", "memory$2", "memory$pair" }, "host refs");
            Eq(secrets.GetFrom("memory", "api.token"), "second", "getfrom memory");
            Eq(secrets.GetFrom("pair", "api.token"), "pair2", "getfrom pair");
        });

        Case("a store name must be a valid tag", () =>
        {
            Eq(Refused(() => new Sekreto(new SekretoOptions
            {
                Providers = new List<object>
                {
                    Spec("kind", "memory", "name", "my store", "values", Spec()),
                },
            }), "a store name with a space"),
                "sekreto: invalid store name: my store", "the message");
        });

        // A provider that refuses its own configuration raises a
        // SekretoError from inside the plugin's define. The spec pins that
        // message byte for byte, so it must come back out of the host as
        // itself - not wrapped as plugin_define_failed, and not as a
        // PluginException.
        Case("a sekreto error raised in define comes back out as itself", () =>
        {
            Eq(Refused(() => new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { Hashicorp.Plugin },
                Providers = new List<object>
                {
                    Spec("kind", "hashicorp", "addr", "http://127.0.0.1:1", "token", "t", "kv", 3),
                },
            }), "a hashicorp kv version that does not exist"),
                "sekreto: hashicorp: unsupported kv version: 3", "the message, byte for byte");
        });

        // ...and any other error is not sekreto's to rewrite: it surfaces as
        // the host reports it, naming the instance and the cause.
        Case("any other error raised in define is the host's report of it", () =>
        {
            Definition broken = Providers.ProviderPlugin("broken",
                spec => throw new InvalidOperationException("boom"));

            Exception err = Caught(() => new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { broken },
                Providers = new List<object> { Spec("kind", "broken") },
            }), "a definition that raises");

            Eq(Plugin.CodeOf(err), "plugin_define_failed", "the host's code");
            True(err.Message.Contains("boom"), "the cause survives: " + err.Message);
            True(!(err is SekretoError), "it is not rewritten as a SekretoError");
        });

        Case("a custom kind is one providerplugin call", () =>
        {
            Definition shouty = Providers.ProviderPlugin("shouty",
                spec => new Shouty(spec.GetValueOrDefault("values") as Dictionary<string, object>));

            var secrets = new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { shouty },
                Providers = new List<object>
                {
                    Spec("kind", "shouty", "values", Spec("API.TOKEN", "loud")),
                },
            });

            Eq(secrets.Get("api.token"), "loud", "get");
            Eq(secrets.Host.List().Keys, new[] { "shouty" }, "host refs");
        });

        // A plugin that names a built-in kind replaces it: that is how a
        // host substitutes an implementation, and never an accident,
        // because the four names are documented.
        Case("a plugin may replace a built-in kind", () =>
        {
            var secrets = new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition>
                {
                    Providers.ProviderPlugin("memory", spec => new Replaced()),
                },
                Providers = new List<object>
                {
                    Spec("kind", "memory", "values", Spec("API_TOKEN", "original")),
                },
            });

            Eq(secrets.Get("api.token"), "replaced", "the substitute answered");
        });

        Case("close tears the chain down and keeps redaction", () =>
        {
            var secrets = new Sekreto(new SekretoOptions
            {
                Providers = new List<object>
                {
                    Spec("kind", "memory", "values", Spec("API_TOKEN", "tok01")),
                },
            });

            Eq(secrets.Get("api.token"), "tok01", "get");

            secrets.Close();

            Eq(secrets.Host.List().Keys, new string[0], "host refs");
            Eq(secrets.Stores(), new string[0], "stores");
            Eq(secrets.TryGet("api.token"), null, "try");
            Eq(secrets.Redact("token=tok01"), "token=[redacted]", "redact");
        });

        // --- what the artifact says ---------------------------------------

        // THE CORE REFERENCES NO PLUGIN. A .NET assembly names every
        // assembly it references in its own metadata, and a type in an
        // assembly that is not named there cannot be resolved at run time -
        // so this is the linking boundary itself, read back, not a
        // convention anyone has to keep.
        Case("the core references no plugin", () =>
        {
            var refs = typeof(Sekreto).Assembly.GetReferencedAssemblies()
                .Select(a => a.Name).OrderBy(n => n, StringComparer.Ordinal).ToList();

            True(!refs.Contains("VoxgigSekretoPlugins"),
                "the core references the plugin assembly: " + Show(refs));

            // ...and none of the three platform assemblies only a plugin
            // needs. A socket, a signature and a child process, in that
            // order.
            foreach (string platform in new[]
            {
                "System.Net.Http", "System.Security.Cryptography", "System.Diagnostics.Process",
            })
            {
                True(!refs.Contains(platform),
                    "the core references " + platform + ": " + Show(refs));
            }

            // The reference points the other way, which is what makes the
            // absence structural: a reference back would be a build cycle.
            var back = typeof(Hashicorp).Assembly.GetReferencedAssemblies()
                .Select(a => a.Name).ToList();

            True(back.Contains("VoxgigSekreto"), "the plugins reference the core: " + Show(back));
        });

        // sigv4 moved with the aws plugin, and that is the sharpest
        // instance of the rule: THE CORE OF NO PORT IMPORTS A HASH
        // FUNCTION.
        Case("sigv4 lives with the aws plugin, not in the core", () =>
        {
            Eq(typeof(Sigv4).Assembly.GetName().Name, "VoxgigSekretoPlugins", "sigv4's assembly");
            Eq(typeof(Sekreto).Assembly.GetName().Name, "VoxgigSekreto", "the core's assembly");

            True(!Directory.EnumerateFiles(
                    Path.GetDirectoryName(Here("src", "Sekreto.csproj")), "*.cs")
                .Any(f => File.ReadAllText(f).Contains("System.Security.Cryptography")),
                "the core's source names a crypto namespace");
        });

        // The full set is built on demand: All() hands back a fresh list
        // every time, so a caller cannot mutate the catalog every other
        // caller will get. It is also the ONLY name that reaches all ten -
        // a consumer that wants one takes one.
        Case("the full set is built on demand", () =>
        {
            List<Definition> first = SekretoPlugins.All();
            List<Definition> second = SekretoPlugins.All();

            True(!ReferenceEquals(first, second), "All() returns a shared list");
            Eq(first.Count, 10, "the full set");
            Eq(Names(first), Names(second), "the same definitions");

            Eq(Names(new List<Definition> { Hashicorp.Plugin }), new[] { "hashicorp" },
                "one plugin is one definition");
        });

        // Python refuses a MODULE handed in where a definition belongs.
        // csharp's type system refuses that at compile time - Plugins is a
        // List<Definition> and a namespace is not a value - so what is left
        // to check is the one non-definition the type does admit.
        Case("a non-definition passed as a plugin is refused", () =>
        {
            Exception err = Caught(() => new Sekreto(new SekretoOptions
            {
                Plugins = new List<Definition> { null },
                Providers = new List<object>(),
            }), "a null plugin");

            // A SEKRETO refusal, not the plugin library's. The message
            // names the fix, and it matches java and python byte for byte.
            Eq(err.GetType().Name, "SekretoError", "refused by sekreto, not the catalog");
            Eq(
                err.Message,
                "sekreto: not a plugin definition: null"
                + " - a plugin is what Support.ProviderPlugin(kind, make) returns",
                "the message names the fix");
        });

        return cases;
    }
}

File: php/test/plugins.php
<?php

/**
 * RUN: php test/plugins.php
 * RUN-SOME: php test/plugins.php thecoreremembersnoplugin
 *
 * THE PLUGIN SEAM, from both sides.
 *
 * Moving the provider kinds that open sockets and spawn processes out of
 * the core made a consumer's PLUGIN LIST load-bearing: a kind nobody
 * passed in is not in the catalog, and a chain naming it is refused. That
 * is the intended behaviour, and it means a consumer can be broken without
 * a single conformance test noticing - the conformance suite passes every
 * plugin, so it can never see a missing one. So the full set is pinned
 * here: it holds every kind, every kind builds, and the CLI passes it.
 *
 * The other half is the boundary itself. PHP has no compiler to erase an
 * unused import and no linker to drop an unreferenced object, so the
 * boundary here is the REQUIRE GRAPH plus the namespace: `src/` is
 * `Voxgig\Sekreto`, `plugins/` is `Voxgig\Sekreto\Plugins`, and nothing in
 * the first requires or names anything in the second. `php
 * test/included.php` reports that graph from a fresh interpreter, which is
 * the only place it can be read - this one has required everything, on
 * purpose.
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Test;

use Voxgig\Plugin\PluginError;
use Voxgig\Sekreto\Provider;
use Voxgig\Sekreto\Sekreto;
use Voxgig\Sekreto\SekretoError;

use function Voxgig\Sekreto\builtins;
use function Voxgig\Sekreto\providerplugin;
use function Voxgig\Sekreto\Plugins\allplugins;
use function Voxgig\Sekreto\Plugins\hashicorp;

require_once __DIR__ . '/../src/Sekreto.php';
require_once __DIR__ . '/../plugins/plugins.php';
require_once __DIR__ . '/../plugins/hashicorp.php';

const PLUGINS = [
    'awsparams', 'awssecrets', 'azuresecrets', 'boru', 'doppler', 'gcpsecrets',
    'hashicorp', 'infisical', 'onepassword', 'secretspec',
];

const EVERY = [
    'awsparams', 'awssecrets', 'azuresecrets', 'boru', 'doppler', 'dotenv',
    'env', 'file', 'gcpsecrets', 'hashicorp', 'infisical', 'memory',
    'onepassword', 'secretspec',
];

$only = $argv[1] ?? null;
$pass = 0;
$fail = 0;

function testcase(string $name, callable $body): void
{
    global $only, $pass, $fail;

    if (null !== $only && $name !== $only) {
        return;
    }

    try {
        $body();
        $pass++;
        echo "ok   - $name\n";
    } catch (\Throwable $err) {
        $fail++;
        echo "FAIL - $name\n     " . $err->getMessage() . "\n";
    }
}

/** @param mixed $want @param mixed $got */
function same($want, $got, string $what = ''): void
{
    if ($want !== $got) {
        throw new \RuntimeException(
            ('' === $what ? '' : $what . ': ') . 'want ' . json_encode($want)
            . ', got ' . json_encode($got)
        );
    }
}

/** @return string the message of the exception the body threw */
function threw(string $class, callable $body): \Throwable
{
    try {
        $body();
    } catch (\Throwable $err) {
        if (!($err instanceof $class)) {
            throw new \RuntimeException(
                'want a ' . $class . ', got a ' . get_class($err) . ': ' . $err->getMessage()
            );
        }
        return $err;
    }

    throw new \RuntimeException('want a ' . $class . ', nothing was thrown');
}

/** Names, sorted, of a list of definitions. @param array<int, mixed> $defs */
function names(array $defs): array
{
    $out = array_map(fn(array $def) => $def['name'], $defs);
    sort($out, SORT_STRING);
    return $out;
}

/**
 * One file's CODE, with every comment stripped.
 *
 * The boundary is what the core executes and names, not what it explains:
 * `Sekreto.php` documents the split in its header, and a grep of the raw
 * text would read its own documentation as a breach.
 */
function codeof(string $file): string
{
    $out = '';

    foreach (token_get_all((string) file_get_contents($file)) as $token) {
        if (is_array($token)) {
            if (T_COMMENT === $token[0] || T_DOC_COMMENT === $token[0]) {
                continue;
            }
            $out .= $token[1];
            continue;
        }

        $out .= $token;
    }

    return $out;
}

/**
 * What requiring one file costs, in a FRESH interpreter.
 *
 * @return array<int, string>
 */
function included(string $target): array
{
    $cmd = escapeshellarg(PHP_BINARY) . ' ' . escapeshellarg(__DIR__ . '/included.php')
        . ' ' . escapeshellarg($target) . ' 2>&1';
    $text = shell_exec($cmd);
    $out = array_values(array_filter(array_map('trim', explode("\n", (string) $text))));
    sort($out, SORT_STRING);
    return $out;
}

// --- the full set --------------------------------------------------------

testcase('thefullsetholdseverykind', function (): void {
    same(PLUGINS, names(allplugins()));
    same(\Voxgig\Sekreto\BUILTIN_KINDS, array_map(fn(array $d) => $d['name'], builtins()));

    $plugin = \Voxgig\Sekreto\PLUGIN_KINDS;
    sort($plugin, SORT_STRING);
    same(PLUGINS, $plugin);
});

// Naming a kind is not enough: a kind can be in the catalog and still fail
// to build. Construction is what the CLI does before any network.
testcase('everykindbuildsfromaspec', function (): void {
    $chain = [];
    foreach (EVERY as $kind) {
        $chain[] = [
            'kind' => $kind, 'addr' => 'http://127.0.0.1:8200', 'token' => 't',
            'dir' => '/tmp', 'file' => '/tmp/.env', 'values' => [],
        ];
    }

    $secrets = new Sekreto(['plugins' => allplugins(), 'providers' => $chain]);

    same(EVERY, $secrets->stores());
    same(EVERY, array_keys($secrets->host->list()));
    same(['live'], array_values(array_unique(array_values($secrets->host->list()))));
});

testcase('theclipassesthefullset', function (): void {
    $src = (string) file_get_contents(__DIR__ . '/../cli/sekreto-cli.php');
    same(true, str_contains($src, "require_once __DIR__ . '/../plugins/plugins.php';"));
    same(true, str_contains($src, "'plugins' => allplugins()"));
});

// --- what a consumer sees ------------------------------------------------

testcase('onepluginisenoughforachainthatnamesonlyit', function (): void {
    $secrets = new Sekreto([
        'plugins' => [hashicorp()],
        'providers' => [
            ['kind' => 'memory', 'values' => ['API_TOKEN' => 'tok01']],
            ['kind' => 'hashicorp', 'name' => 'prod',
             'addr' => 'https://vault.example.com', 'token' => 't'],
        ],
    ]);

    same(['memory', 'prod'], $secrets->stores());
    same(['memory', 'hashicorp:https://vault.example.com/secret'], $secrets->sources());
    same('tok01', $secrets->get('api.token'));

    // The plugin host is what the chain is made of, and it reads like the
    // chain: the kind, or kind$store for a named store.
    same(['hashicorp$prod' => 'live', 'memory' => 'live'], $secrets->host->list());
    same(['dotenv', 'env', 'file', 'hashicorp', 'memory'], $secrets->catalog->names());
});

testcase('akindthatwasnotpassedinisrefusednamingthefix', function (): void {
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'plugins' => [hashicorp()],
        'providers' => [['kind' => 'doppler', 'token' => 't']],
    ]));
    same(
        'sekreto: unknown provider kind: doppler'
        . ' (available: dotenv, env, file, hashicorp, memory)'
        . ' - doppler is a sekreto plugin, not built in: pass it in the plugins option',
        $err->getMessage()
    );

    // A kind nobody ships is a typo, and gets no such hint.
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'providers' => [['kind' => 'vualt']],
    ]));
    same(
        'sekreto: unknown provider kind: vualt (available: dotenv, env, file, memory)',
        $err->getMessage()
    );
});

// Two providers MAY share a store name - a directed read walks both, and
// the spec pins it - but an instance ref may not, so the second gets a
// numbered tag from the host and keeps its store name.
testcase('arepeatedstorenamekeepsthestoreandnumberstheinstance', function (): void {
    $secrets = new Sekreto(['providers' => [
        ['kind' => 'memory', 'values' => []],
        ['kind' => 'memory', 'values' => ['API_TOKEN' => 'second']],
        ['kind' => 'memory', 'name' => 'pair', 'values' => []],
        ['kind' => 'memory', 'name' => 'pair', 'values' => ['API_TOKEN' => 'pair2']],
    ]]);

    same(['memory', 'pair'], $secrets->stores());
    same(
        ['memory', 'memory$1', 'memory$2', 'memory$pair'],
        array_keys($secrets->host->list())
    );
    same('second', $secrets->getfrom('memory', 'api.token'));
    same('pair2', $secrets->getfrom('pair', 'api.token'));
});

testcase('astorenamemustbeavalidtag', function (): void {
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'providers' => [['kind' => 'memory', 'name' => 'my store', 'values' => []]],
    ]));
    same('sekreto: invalid store name: my store', $err->getMessage());
});

// A provider that refuses its own configuration throws a SekretoError from
// inside the plugin's `define`. The spec pins that message byte for byte,
// so it must come back out of the host as itself - not wrapped as
// plugin_define_failed, and not as a PluginError.
testcase('asekretoerrorthrownindefinecomesbackoutasitself', function (): void {
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'plugins' => [hashicorp()],
        'providers' => [['kind' => 'hashicorp', 'addr' => 'http://127.0.0.1:1',
                         'token' => 't', 'kv' => 3]],
    ]));
    same('sekreto: hashicorp: unsupported kv version: 3', $err->getMessage());
});

// ...and any other error is not sekreto's to rewrite: it surfaces as the
// host reports it, naming the instance and the cause.
testcase('anyothererrorthrownindefineisthehostsreportofit', function (): void {
    $broken = providerplugin('broken', function (array $spec): Provider {
        throw new \TypeError('boom');
    });

    $err = threw(PluginError::class, fn() => new Sekreto([
        'plugins' => [$broken],
        'providers' => [['kind' => 'broken']],
    ]));
    same('plugin_define_failed', $err->code);
    same(true, str_contains($err->getMessage(), 'boom'));
});

testcase('acustomkindisoneproviderplugincall', function (): void {
    $shouty = providerplugin('shouty', fn(array $spec) => new class ($spec['values'] ?? []) implements Provider {
        /** @param array<string, string> $values */
        public function __construct(private array $values)
        {
        }

        public function lookup(string $name): ?string
        {
            return $this->values[strtoupper($name)] ?? null;
        }

        public function describe(): string
        {
            return 'shouty';
        }
    });

    $secrets = new Sekreto([
        'plugins' => [$shouty],
        'providers' => [['kind' => 'shouty', 'values' => ['API.TOKEN' => 'loud']]],
    ]);

    same('loud', $secrets->get('api.token'));
    same(['shouty' => 'live'], $secrets->host->list());
});

// A plugin that names a built-in kind replaces it: that is how a host
// substitutes an implementation, and never an accident, because the four
// names are documented.
testcase('apluginmayreplaceabuiltinkind', function (): void {
    $replaced = providerplugin('memory', fn(array $spec) => new class implements Provider {
        public function lookup(string $name): ?string
        {
            return 'replaced';
        }

        public function describe(): string
        {
            return 'memory';
        }
    });

    $secrets = new Sekreto([
        'plugins' => [$replaced],
        'providers' => [['kind' => 'memory', 'values' => ['API_TOKEN' => 'original']]],
    ]);

    same('replaced', $secrets->get('api.token'));
});

testcase('closetearsthechaindownandkeepsredaction', function (): void {
    $secrets = new Sekreto([
        'providers' => [['kind' => 'memory', 'values' => ['API_TOKEN' => 'tok01']]],
    ]);
    same('tok01', $secrets->get('api.token'));

    $secrets->close();

    same([], $secrets->host->list());
    same([], $secrets->stores());
    same(null, $secrets->try('api.token'));
    same('token=[redacted]', $secrets->redact('token=tok01'));
});

// --- the boundary --------------------------------------------------------

// The core requires no plugin: requiring src/Sekreto.php brings in the
// chain, the built-ins, the address checks and voxgig/plugin, and not one
// file under plugins/.
testcase('thecorerequiresnoplugin', function (): void {
    same(
        ['src/Addr.php', 'src/Plugin.php', 'src/Providers.php', 'src/Sekreto.php'],
        included('src/Sekreto.php')
    );

    // ...and it names none of them either. PHP erases nothing, so the
    // namespace is the second half of the boundary: CODE in the core that
    // never writes `Voxgig\Sekreto\Plugins` and never requires a path under
    // plugins/ cannot reach a plugin by accident. Comments are stripped
    // first - this file's own header documents the split, and prose about
    // the boundary is not a breach of it.
    foreach (glob(__DIR__ . '/../src/*.php') as $file) {
        $code = codeof($file);
        same(false, str_contains($code, 'Sekreto\\Plugins'), basename($file));
        same(false, str_contains($code, '/plugins/'), basename($file));

        // What the plugins hold, the core does not: no socket, no
        // signature, no child process. `file_get_contents` stays - reading
        // a local file is what makes a kind built in.
        foreach (['proc_open', 'stream_context_create', 'fsockopen', 'curl_init',
                  'hash_hmac', 'hash(', 'openssl_'] as $call) {
            same(false, str_contains($code, $call), basename($file) . ' uses ' . $call);
        }
    }
});

// ...and one plugin requires only itself, plus the core and the shared
// HTTP helper it speaks through. Requiring one kind must not drag in AWS
// request signing and the two child-process kinds behind it.
testcase('onepluginrequiresonlyitself', function (): void {
    same(
        ['plugins/hashicorp.php', 'plugins/httpjson.php', 'src/Addr.php',
         'src/Plugin.php', 'src/Providers.php', 'src/Sekreto.php'],
        included('plugins/hashicorp.php')
    );
});

// The full set is built on demand, and reaching it requires everything.
// The requires live inside `allplugins()` for exactly this reason: at file
// scope, naming the list would load all ten as a side effect.
testcase('thefullsetisbuiltondemand', function (): void {
    same(['plugins/plugins.php'], included('plugins/plugins.php'));

    $after = included('test/allplugins.php');
    foreach (['hashicorp', 'boru', 'aws', 'sigv4', 'gcpsecrets', 'azuresecrets',
              'onepassword', 'doppler', 'infisical', 'secretspec', 'httpjson',
              'runcmd'] as $mod) {
        same(true, in_array('plugins/' . $mod . '.php', $after, true), $mod);
    }
});

// A plugin file holds a FUNCTION that builds the definition, because a
// definition holds closures and PHP has no constant that can. Passing the
// function instead of calling it is the mistake that shape makes easy, so
// it is refused by name, saying what to do instead.
testcase('afunctionpassedasapluginisrefused', function (): void {
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'plugins' => [hashicorp(...)],
        'providers' => [],
    ]));
    same(
        'sekreto: not a plugin definition: the function Voxgig\\Sekreto\\Plugins\\hashicorp'
        . ' - call it for the definition it holds',
        $err->getMessage()
    );

    // And anything else that is not a definition is refused too.
    $err = threw(SekretoError::class, fn() => new Sekreto([
        'plugins' => ['hashicorp'],
        'providers' => [],
    ]));
    same("sekreto: not a plugin definition: 'hashicorp'", $err->getMessage());
});

echo "\n$pass passed, $fail failed\n";
exit(0 === $fail ? 0 : 1);
